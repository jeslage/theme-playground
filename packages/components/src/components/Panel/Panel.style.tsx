import styled from 'styled-components';

const StyledPanel = styled.div`
  height: calc(100% - 2rem);
  margin: 0;
  display: flex;
  flex-direction: column;
  background: ${props => props.theme.backgroundDark};
  color: ${props => props.theme.text};

  .panel__content {
    position: relative;
    flex-grow: 2;
  }

  .panel__themes {
    display: flex;
    justify-content: center;
    padding: 1rem;
    min-height: 50px;
    background: ${props => props.theme.backgroundLight};
  }
`;

export default StyledPanel;
