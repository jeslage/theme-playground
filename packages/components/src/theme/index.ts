export const darkTheme = {
  base: 'dark',
  backgroundDark: '#2f2f2f',
  backgroundLight: '#333',
  border: '#999999',
  text: '#FFFFFF'
};

export const lightTheme = {
  base: 'light',
  backgroundDark: '#f6f9fc',
  backgroundLight: '#fff',
  border: '#666666',
  text: '#333333'
};

export const getUiTheme = theme => {
  if (theme === undefined) return darkTheme;

  if (typeof theme === 'string') {
    if (theme === 'dark') {
      return darkTheme;
    } else if (theme === 'light') {
      return lightTheme;
    }

    throw Error(
      `Theme "${theme}" is not available. Please use "light" or "dark", or provide a custom theme object.`
    );
  }

  if (theme.base && theme.base === 'light') {
    return {
      ...lightTheme,
      ...theme
    };
  }

  return {
    ...darkTheme,
    ...theme
  };
};
