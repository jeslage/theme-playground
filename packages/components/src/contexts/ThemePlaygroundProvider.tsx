import * as React from 'react';
import { ThemeProvider } from 'styled-components';

import { updateValueBasedOnPath } from '../helper';
import {
  Theme,
  ThemesArray,
  ThemeObject,
  ConfigProps,
  OptionsType,
  Overrides,
  ThemePlaygroundContextProps,
  ThemePlaygroundProviderProps
} from '../definitions';

import buildThemeComponents from '../helper/buildThemeComponents';
import { getUiTheme } from '../theme';

const defaultConfig = {
  labelFormat: 'startCase',
  showCode: true
};

const defaultProps = {
  themes: [],
  themeComponents: {},
  activeTheme: { name: '__default', theme: {} },
  overrides: {},
  config: defaultConfig,
  isLoading: false,
  updateIsLoading: () => {},
  updateTheme: () => {},
  updateActiveTheme: () => {},
  getInitialOptions: () => {}
};

export const ThemePlaygroundContext = React.createContext<
  ThemePlaygroundContextProps
>(defaultProps);

const ThemePlaygroundProvider: React.FC<ThemePlaygroundProviderProps> = ({
  options,
  uiTheme,
  children
}) => {
  const [isLoading, setIsLoading] = React.useState<boolean>(false);

  const [themeComponents, setThemeComponents] = React.useState({});
  const [themes, setThemes] = React.useState<ThemesArray>(
    Array.isArray(options.theme) ? options.theme : []
  );

  const getActiveTheme = () => {
    if (options.theme) {
      if (Array.isArray(options.theme)) {
        return { ...options.theme[0] };
      }
      return { name: '__default', theme: options.theme };
    }

    return { name: '', theme: {} };
  };

  const [activeTheme, setActiveTheme] = React.useState<ThemeObject>(
    getActiveTheme()
  );

  const [overrides, setOverrides] = React.useState<Overrides>(
    options.overrides || {}
  );
  const [config, setConfig] = React.useState<ConfigProps>({
    ...options.config,
    ...defaultConfig
  });

  const updateThemeComponents = (theme: Theme, overrides: Overrides) => {
    const components: { [key: string]: any } = {};

    if (Array.isArray(theme)) {
      theme.forEach(({ name, theme }) => {
        components[name] = buildThemeComponents(theme, overrides);
      });
    } else {
      components.__default = buildThemeComponents(theme, overrides);
    }

    setThemeComponents(components);
  };

  React.useEffect(
    () => updateThemeComponents(options.theme || {}, options.overrides || {}),
    []
  );

  const getInitialOptions = (options: OptionsType) => {
    const { theme, overrides, config } = options;

    updateThemeComponents(theme, overrides || {});

    if (Array.isArray(theme)) {
      setThemes(theme);
      setActiveTheme({ ...theme[0] });
    } else {
      setActiveTheme({ name: '__default', theme });
    }

    if (overrides) setOverrides(overrides);
    if (config) setConfig(prev => ({ ...prev, ...config }));
  };

  const updateTheme = React.useCallback(
    (path: string, value: any) => {
      const { theme, name } = activeTheme;

      // Update theme object value based on path and set active theme state
      const newTheme: Theme = theme;
      updateValueBasedOnPath(path, value, newTheme);
      setActiveTheme({ name, theme: newTheme });

      // Set new theme components state
      setThemeComponents(prev => ({
        ...prev,
        [name]: {
          ...prev[name],
          [path]: { type: prev[name][path].type, value }
        }
      }));
    },
    [activeTheme]
  );

  const updateActiveTheme = React.useCallback(
    ({ name, theme }: ThemeObject) => {
      setActiveTheme({ name, theme });
    },
    [activeTheme.theme]
  );

  const providerValue: ThemePlaygroundContextProps = {
    activeTheme,
    themes,
    themeComponents,
    config,
    overrides,
    updateTheme,
    updateActiveTheme,
    getInitialOptions,
    isLoading,
    updateIsLoading: bool => setIsLoading(bool)
  };

  return (
    <ThemePlaygroundContext.Provider value={providerValue}>
      <ThemeProvider theme={getUiTheme(uiTheme)}>
        {children && children(activeTheme || {})}
      </ThemeProvider>
    </ThemePlaygroundContext.Provider>
  );
};

export const ThemePlaygroundConsumer = ThemePlaygroundContext.Consumer;

export default ThemePlaygroundProvider;
