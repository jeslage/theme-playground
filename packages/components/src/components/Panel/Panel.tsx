import * as React from 'react';

import { ThemePlaygroundContext } from '../../contexts/ThemePlaygroundProvider';

import Code from '../Code/Code';
import SettingsItem from '../SettingsItem/SettingsItem';
import RadioGroup from '../RadioGroup/RadioGroup';

import StyledPanel from './Panel.style';

const Panel = ({ className }: { className?: string | null }) => {
  const {
    themes,
    activeTheme: { theme, name },
    updateActiveTheme,
    config
  } = React.useContext(ThemePlaygroundContext);

  return (
    <StyledPanel className={className}>
      {Object.entries(theme).length > 0 && (
        <>
          <div className="panel__content">
            {themes.length > 1 && (
              <div className="panel__themes">
                <RadioGroup
                  label="Active Theme"
                  name="themes"
                  value={name}
                  onChange={val =>
                    updateActiveTheme(themes.filter(t => t.name === val)[0])
                  }
                  options={themes.map(t => ({ label: t.name, value: t.name }))}
                />
              </div>
            )}
            <SettingsItem />
            {config.showCode && <Code value={theme} />}
          </div>
        </>
      )}
    </StyledPanel>
  );
};

export default Panel;
