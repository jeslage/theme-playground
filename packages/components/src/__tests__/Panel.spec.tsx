import 'jest';
import * as React from 'react';
import { shallow } from 'enzyme';
import Panel from '../components/Panel/Panel';
import StyledPanel from '../components/Panel/Panel.style';
import { darkTheme } from '../theme';
import ThemePlaygroundProvider from '../contexts/ThemePlaygroundProvider';

describe('Panel', () => {
  it('Panel should render correctly', () => {
    const component = shallow(
      <ThemePlaygroundProvider options={{ theme: { color: 'red' } }}>
        {() => <Panel className="panel__classname" />}
      </ThemePlaygroundProvider>
    );

    expect(component).toMatchSnapshot();
  });

  it('StyledPanel should render correctly', () => {
    const component = shallow(<StyledPanel theme={{ colors: darkTheme }} />);

    expect(component).toMatchSnapshot();
  });
});
