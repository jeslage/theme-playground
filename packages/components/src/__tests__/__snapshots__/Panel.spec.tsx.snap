// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Panel Panel should render correctly 1`] = `
<ContextProvider
  value={
    Object {
      "activeTheme": Object {
        "name": "__default",
        "theme": Object {
          "color": "red",
        },
      },
      "config": Object {
        "labelFormat": "startCase",
        "showCode": true,
      },
      "getInitialOptions": [Function],
      "isLoading": false,
      "overrides": Object {},
      "themeComponents": Object {},
      "themes": Array [],
      "updateActiveTheme": [Function],
      "updateIsLoading": [Function],
      "updateTheme": [Function],
    }
  }
>
  <Panel
    className="panel__classname"
  />
</ContextProvider>
`;

exports[`Panel StyledPanel should render correctly 1`] = `
<div
  className="sc-fzplWN hCHpGs"
/>
`;
